File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/item_detail/bodies/MemoryBody.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongMethod", "TooManyFunctions", "UnusedPrivateMember")

package app.pantopus.android.ui.screens.mailbox.item_detail.bodies

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.data.api.models.mailbox.v2.MemoryDetailDto
import app.pantopus.android.data.api.models.mailbox.v2.MemoryElfBullet
import app.pantopus.android.data.api.models.mailbox.v2.MemoryElfContent
import app.pantopus.android.data.api.models.mailbox.v2.MemoryFact
import app.pantopus.android.data.api.models.mailbox.v2.MemoryVaultCrumb
import app.pantopus.android.data.api.models.mailbox.v2.MemoryVaultInfo
import app.pantopus.android.ui.screens.mailbox.item_detail.bodies.components.PolaroidFrame
import app.pantopus.android.ui.screens.mailbox.item_detail.bodies.components.StationeryCard
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * Concrete body for the Memory mailbox category (A17.7). A keepsake
 * delivery: a serif title + reference, the polaroid photograph, the
 * handwritten note, the "Pantopus surfaced this" elf, and a contextual
 * summary that swaps from the facts grid (fresh) to the vault-location
 * card once kept. The shell owns the accent strip, sender block, trust
 * pill, and the sticky Save-to-Vault / Share shelf.
 */
@Composable
fun MemoryBody(
    memory: MemoryDetailDto,
    isSaved: Boolean,
    modifier: Modifier = Modifier,
    onOpenThread: (() -> Unit)? = null,
    onOpenVault: (() -> Unit)? = null,
) {
    Column(
        modifier = modifier.fillMaxWidth().testTag("memoryBody"),
        verticalArrangement = Arrangement.spacedBy(Spacing.s4),
    ) {
        TitleBlock(memory, Modifier.padding(horizontal = Spacing.s4))

        if (isSaved) {
            SavedBanner(Modifier.padding(horizontal = Spacing.s4))
        }

        PolaroidFrame(
            photoUrl = memory.photoUrl,
            caption = memory.photoCaption,
            label = memory.photoLabel,
            modifier = Modifier.padding(horizontal = Spacing.s4),
        )

        StationeryCard(
            eyebrow = "The note",
            paragraphs = memory.note,
            signature = memory.noteSignature,
            modifier = Modifier.padding(horizontal = Spacing.s4),
        )

        MemoryElfCard(
            content = if (isSaved) memory.elfSaved else memory.elfFresh,
            modifier = Modifier.padding(horizontal = Spacing.s4),
        )

        if (isSaved) {
            MemoryVaultCard(
                vault = memory.vault,
                onOpenVault = onOpenVault,
                modifier = Modifier.padding(horizontal = Spacing.s4),
            )
        } else {
            MemoryFactsCard(
                facts = memory.facts,
                onOpenThread = onOpenThread,
                modifier = Modifier.padding(horizontal = Spacing.s4),
            )
        }
    }
}

@Composable
private fun TitleBlock(
    memory: MemoryDetailDto,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(Spacing.s1)) {
        Text(
            text = memory.title,
            fontFamily = FontFamily.Serif,
            fontSize = 22.sp,
            fontWeight = FontWeight.SemiBold,
            color = PantopusColors.appText,
            modifier = Modifier.semantics { heading() },
        )
        Text(
            text = memory.reference,
            fontFamily = FontFamily.Monospace,
            fontSize = 11.sp,
            color = PantopusColors.appTextSecondary,
        )
    }
}

@Composable
private fun SavedBanner(modifier: Modifier = Modifier) {
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.md))
                .background(PantopusColors.successBg)
                .border(1.dp, PantopusColors.successLight, RoundedCornerShape(Radii.md))
                .padding(horizontal = Spacing.s3, vertical = Spacing.s2)
                .semantics { contentDescription = "Kept in your Vault, only you can see it" },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Box(
            modifier =
                Modifier
                    .size(20.dp)
                    .clip(CircleShape)
                    .background(PantopusColors.success),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(
                icon = PantopusIcon.Heart,
                contentDescription = null,
                size = 11.dp,
                tint = PantopusColors.appTextInverse,
            )
        }
        Text(
            text =
                buildAnnotatedString {
                    withStyle(SpanStyle(fontWeight = FontWeight.Bold, color = PantopusColors.success)) {
                        append("Kept in your Vault")
                    }
                    withStyle(SpanStyle(color = PantopusColors.appTextSecondary)) {
                        append(" · only you can see it")
                    }
                },
            fontSize = 12.sp,
        )
    }
}

@Composable
private fun MemoryElfCard(
    content: MemoryElfContent,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.xl))
                .background(PantopusColors.primary50)
                .border(1.dp, PantopusColors.primary200, RoundedCornerShape(Radii.xl))
                .padding(Spacing.s3),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            Box(
                modifier =
                    Modifier
                        .size(24.dp)
                        .clip(RoundedCornerShape(Radii.md))
                        .background(PantopusColors.primary600),
                contentAlignment = Alignment.Center,
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.Sparkles,
                    contentDescription = null,
                    size = 13.dp,
                    tint = PantopusColors.appTextInverse,
                )
            }
            Text(
                text = content.headline,
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                color = PantopusColors.primary800,
            )
        }

        Text(
            text = content.summary,
            fontSize = 13.sp,
            lineHeight = 18.sp,
            color = PantopusColors.primary900,
        )

        Column(verticalArrangement = Arrangement.spacedBy(Spacing.s2)) {
            content.bullets.forEach { bullet -> BulletRow(bullet) }
        }
    }
}

@Composable
private fun BulletRow(bullet: MemoryElfBullet) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        verticalAlignment = Alignment.Top,
        modifier = Modifier.semantics { contentDescription = "${bullet.label}. ${bullet.text}" },
    ) {
        Box(
            modifier =
                Modifier
                    .size(18.dp)
                    .clip(RoundedCornerShape(Radii.xs))
                    .background(PantopusColors.appSurface)
                    .border(1.dp, PantopusColors.primary200, RoundedCornerShape(Radii.xs)),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(
                icon = bulletIcon(bullet.glyph),
                contentDescription = null,
                size = 10.dp,
                tint = PantopusColors.primary700,
            )
        }
        Text(
            text =
                buildAnnotatedString {
                    withStyle(SpanStyle(fontWeight = FontWeight.Bold, color = PantopusColors.appText)) {
                        append(bullet.label)
                    }
                    withStyle(SpanStyle(color = PantopusColors.appTextSecondary)) {
                        append(" — ${bullet.text}")
                    }
                },
            fontSize = 12.sp,
        )
    }
}

@Composable
private fun MemoryFactsCard(
    facts: List<MemoryFact>,
    onOpenThread: (() -> Unit)?,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.xl))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.xl)),
    ) {
        Text(
            text = "The story behind it",
            style = PantopusTextStyle.overline,
            color = PantopusColors.appTextSecondary,
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
        )
        HorizontalDivider(color = PantopusColors.appBorderSubtle)

        facts.forEachIndexed { index, fact ->
            FactEntry(fact = fact, onOpenThread = onOpenThread)
            if (index < facts.lastIndex) {
                HorizontalDivider(color = PantopusColors.appBorderSubtle)
            }
        }
    }
}

@Composable
private fun FactEntry(
    fact: MemoryFact,
    onOpenThread: (() -> Unit)?,
) {
    val onThreadClick = if (fact.linkHint != null) onOpenThread else null
    val rowModifier =
        if (onThreadClick != null) {
            Modifier
                .clickable(onClick = onThreadClick)
                .testTag("memoryBody.openThread")
                .semantics { contentDescription = "Open the original Pulse thread" }
        } else {
            Modifier
        }
    FactRow(fact = fact, showHint = onThreadClick != null, modifier = rowModifier)
}

@Composable
private fun FactRow(
    fact: MemoryFact,
    showHint: Boolean,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .heightIn(min = 44.dp)
                .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
        horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
        verticalAlignment = Alignment.Top,
    ) {
        Box(
            modifier =
                Modifier
                    .size(24.dp)
                    .clip(RoundedCornerShape(Radii.sm))
                    .background(PantopusColors.warningBg),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(
                icon = factIcon(fact.kind),
                contentDescription = null,
                size = 13.dp,
                tint = PantopusColors.warning,
            )
        }
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(1.dp),
        ) {
            Text(
                text = fact.label,
                fontSize = 11.sp,
                fontWeight = FontWeight.SemiBold,
                color = PantopusColors.appTextSecondary,
            )
            Text(
                text = fact.value,
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                color = PantopusColors.appText,
            )
            if (showHint && fact.linkHint != null) {
                Text(
                    text = fact.linkHint,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.primary600,
                )
            }
        }
    }
}

@Composable
private fun MemoryVaultCard(
    vault: MemoryVaultInfo,
    onOpenVault: (() -> Unit)?,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.xl))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.xl))
                .padding(Spacing.s4),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Text(
            text = "Filed in your Vault",
            style = PantopusTextStyle.overline,
            color = PantopusColors.appTextSecondary,
        )
        Breadcrumb(vault)
        VaultStats(vault)
        if (onOpenVault != null) {
            OpenVaultButton(onOpenVault)
        }
    }
}

@Composable
private fun Breadcrumb(vault: MemoryVaultInfo) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.md))
                .background(PantopusColors.appSurfaceSunken)
                .horizontalScroll(rememberScrollState())
                .padding(horizontal = Spacing.s2, vertical = Spacing.s2)
                .semantics {
                    contentDescription = "Path: " + vault.trail.joinToString(", ") { it.label }
                },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        vault.trail.forEachIndexed { index, crumb ->
            CrumbChip(crumb)
            if (index < vault.trail.lastIndex) {
                PantopusIconImage(
                    icon = PantopusIcon.ChevronRight,
                    contentDescription = null,
                    size = 11.dp,
                    tint = PantopusColors.appTextMuted,
                )
            }
        }
    }
}

@Composable
private fun CrumbChip(crumb: MemoryVaultCrumb) {
    val foreground = if (crumb.isCurrent) PantopusColors.appText else PantopusColors.appTextStrong
    val chipModifier =
        if (crumb.isCurrent) {
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.pill))
        } else {
            Modifier.clip(RoundedCornerShape(Radii.pill))
        }
    Row(
        modifier = chipModifier.padding(horizontal = Spacing.s2, vertical = Spacing.s1),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        PantopusIconImage(
            icon = crumbIcon(crumb.glyph),
            contentDescription = null,
            size = 11.dp,
            tint = foreground,
        )
        Text(
            text = crumb.label,
            fontSize = 12.sp,
            fontWeight = if (crumb.isCurrent) FontWeight.Bold else FontWeight.SemiBold,
            color = foreground,
        )
    }
}

@Composable
private fun VaultStats(vault: MemoryVaultInfo) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.md))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorderSubtle, RoundedCornerShape(Radii.md))
                .padding(vertical = Spacing.s2),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        vault.stats.forEachIndexed { index, stat ->
            if (index > 0) {
                Box(
                    modifier =
                        Modifier
                            .width(1.dp)
                            .height(28.dp)
                            .background(PantopusColors.appBorderSubtle),
                )
            }
            Column(
                modifier = Modifier.weight(1f),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(2.dp),
            ) {
                Text(
                    text = stat.value,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.ExtraBold,
                    color = PantopusColors.appText,
                    maxLines = 1,
                )
                Text(
                    text = stat.label,
                    fontSize = 10.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.appTextSecondary,
                )
            }
        }
    }
}

@Composable
private fun OpenVaultButton(onOpenVault: () -> Unit) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .heightIn(min = 44.dp)
                .clip(RoundedCornerShape(Radii.md))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.md))
                .clickable(onClick = onOpenVault)
                .testTag("memoryBody.openVault")
                .semantics { contentDescription = "Open Vault, Memories" }
                .padding(Spacing.s2),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = "Open Vault › Memories",
            fontSize = 13.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.appText,
        )
        Spacer(Modifier.width(Spacing.s1))
        PantopusIconImage(
            icon = PantopusIcon.ArrowRight,
            contentDescription = null,
            size = 13.dp,
            tint = PantopusColors.appText,
        )
    }
}

private fun factIcon(kind: MemoryFact.Kind): PantopusIcon =
    when (kind) {
        MemoryFact.Kind.Anniversary -> PantopusIcon.Calendar
        MemoryFact.Kind.PulseThread -> PantopusIcon.MessageSquare
        MemoryFact.Kind.Location -> PantopusIcon.MapPin
        MemoryFact.Kind.Others -> PantopusIcon.Users
    }

private fun bulletIcon(glyph: MemoryElfBullet.Glyph): PantopusIcon =
    when (glyph) {
        MemoryElfBullet.Glyph.Calendar -> PantopusIcon.Calendar
        MemoryElfBullet.Glyph.Image -> PantopusIcon.Image
        MemoryElfBullet.Glyph.ShieldCheck -> PantopusIcon.ShieldCheck
        MemoryElfBullet.Glyph.Archive -> PantopusIcon.Archive
        MemoryElfBullet.Glyph.EyeOff -> PantopusIcon.EyeOff
        MemoryElfBullet.Glyph.Bell -> PantopusIcon.Bell
    }

private fun crumbIcon(glyph: MemoryVaultCrumb.Glyph): PantopusIcon =
    when (glyph) {
        MemoryVaultCrumb.Glyph.Inbox -> PantopusIcon.Inbox
        MemoryVaultCrumb.Glyph.Archive -> PantopusIcon.Archive
        MemoryVaultCrumb.Glyph.Heart -> PantopusIcon.Heart
        MemoryVaultCrumb.Glyph.Calendar -> PantopusIcon.Calendar
    }

@Preview(showBackground = true, widthDp = 360, heightDp = 1600)
@Composable
private fun MemoryBodyFreshPreview() {
    Box(modifier = Modifier.background(PantopusColors.appBg).padding(vertical = Spacing.s4)) {
        MemoryBody(memory = MemorySampleData.memory, isSaved = false, onOpenThread = {})
    }
}

@Preview(showBackground = true, widthDp = 360, heightDp = 1600)
@Composable
private fun MemoryBodySavedPreview() {
    Box(modifier = Modifier.background(PantopusColors.appBg).padding(vertical = Spacing.s4)) {
        MemoryBody(memory = MemorySampleData.savedMemory, isSaved = true, onOpenVault = {})
    }
}

File: frontend/apps/ios/Pantopus/Features/Mailbox/ItemDetail/Bodies/MemoryBody.swift
//
//  MemoryBody.swift
//  Pantopus
//
//  Concrete body for the Memory mailbox category (A17.7). A keepsake
//  delivery: a serif title + reference, the polaroid photograph, the
//  handwritten note, the "Pantopus surfaced this" elf, and a contextual
//  summary that swaps from the facts grid (fresh) to the vault-location
//  card once kept. The shell owns the accent strip, sender block, trust
//  pill, and the sticky Save-to-Vault / Share shelf.
//
// swiftlint:disable multiple_closures_with_trailing_closure

import SwiftUI

@MainActor
public struct MemoryBody: View {
    private let memory: MemoryDetailDTO
    private let isSaved: Bool
    private let onOpenThread: (@MainActor () -> Void)?
    private let onOpenVault: (@MainActor () -> Void)?

    public init(
        memory: MemoryDetailDTO,
        isSaved: Bool,
        onOpenThread: (@MainActor () -> Void)? = nil,
        onOpenVault: (@MainActor () -> Void)? = nil
    ) {
        self.memory = memory
        self.isSaved = isSaved
        self.onOpenThread = onOpenThread
        self.onOpenVault = onOpenVault
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s4) {
            titleBlock
                .padding(.horizontal, Spacing.s4)

            if isSaved {
                MemorySavedBanner()
                    .padding(.horizontal, Spacing.s4)
            }

            PolaroidFrame(
                imageURL: memory.photoURL,
                caption: memory.photoCaption,
                label: memory.photoLabel
            )
            .padding(.horizontal, Spacing.s4)

            StationeryCard(
                eyebrow: "The note",
                paragraphs: memory.note,
                signature: memory.noteSignature
            )
            .padding(.horizontal, Spacing.s4)

            MemoryElfCard(content: isSaved ? memory.elfSaved : memory.elfFresh)
                .padding(.horizontal, Spacing.s4)

            if isSaved {
                MemoryVaultCard(vault: memory.vault, onOpenVault: onOpenVault)
                    .padding(.horizontal, Spacing.s4)
            } else {
                MemoryFactsCard(facts: memory.facts, onOpenThread: onOpenThread)
                    .padding(.horizontal, Spacing.s4)
            }
        }
        .accessibilityIdentifier("memoryBody")
    }

    private var titleBlock: some View {
        VStack(alignment: .leading, spacing: Spacing.s1) {
            Text(memory.title)
                .font(.system(size: 22, weight: .semibold, design: .serif))
                .foregroundStyle(Theme.Color.appText)
                .fixedSize(horizontal: false, vertical: true)
                .accessibilityAddTraits(.isHeader)
            Text(memory.reference)
                .font(.system(size: 11, design: .monospaced))
                .foregroundStyle(Theme.Color.appTextSecondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

// MARK: - Saved banner

private struct MemorySavedBanner: View {
    var body: some View {
        HStack(spacing: Spacing.s2) {
            ZStack {
                Circle().fill(Theme.Color.success)
                Icon(.heart, size: 11, color: Theme.Color.appTextInverse)
            }
            .frame(width: 20, height: 20)

            Text("Kept in your Vault").fontWeight(.bold).foregroundStyle(Theme.Color.success)
                + Text(" · only you can see it").foregroundStyle(Theme.Color.appTextSecondary)
        }
        .font(Theme.Font.caption)
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, Spacing.s2)
        .background(Theme.Color.successBg)
        .clipShape(RoundedRectangle(cornerRadius: Radii.md))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.md)
                .stroke(Theme.Color.successLight, lineWidth: 1)
        )
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Kept in your Vault, only you can see it")
    }
}

// MARK: - Elf card

private struct MemoryElfCard: View {
    let content: MemoryElfContent

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            HStack(spacing: Spacing.s2) {
                ZStack {
                    RoundedRectangle(cornerRadius: Radii.md).fill(Theme.Color.primary600)
                    Icon(.sparkles, size: 13, color: Theme.Color.appTextInverse)
                }
                .frame(width: 24, height: 24)
                Text(content.headline)
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.Color.primary800)
            }

            Text(content.summary)
                .font(.system(size: 13))
                .foregroundStyle(Theme.Color.primary900)
                .lineSpacing(3)
                .fixedSize(horizontal: false, vertical: true)

            VStack(alignment: .leading, spacing: Spacing.s2) {
                ForEach(content.bullets) { bullet in
                    bulletRow(bullet)
                }
            }
            .padding(.top, Spacing.s1)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Spacing.s3)
        .background(Theme.Color.primary50)
        .clipShape(RoundedRectangle(cornerRadius: Radii.xl))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.xl)
                .stroke(Theme.Color.primary200, lineWidth: 1)
        )
    }

    private func bulletRow(_ bullet: MemoryElfBullet) -> some View {
        HStack(alignment: .top, spacing: Spacing.s2) {
            ZStack {
                RoundedRectangle(cornerRadius: Radii.xs).fill(Theme.Color.appSurface)
                RoundedRectangle(cornerRadius: Radii.xs).stroke(Theme.Color.primary200, lineWidth: 1)
                Icon(icon(for: bullet.glyph), size: 10, color: Theme.Color.primary700)
            }
            .frame(width: 18, height: 18)

            (Text(bullet.label).fontWeight(.bold).foregroundStyle(Theme.Color.appText)
                + Text(" — \(bullet.text)").foregroundStyle(Theme.Color.appTextSecondary))
                .pantopusTextStyle(.caption)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(bullet.label). \(bullet.text)")
    }

    private func icon(for glyph: MemoryElfBullet.Glyph) -> PantopusIcon {
        switch glyph {
        case .calendar: .calendar
        case .image: .image
        case .shieldCheck: .shieldCheck
        case .archive: .archive
        case .eyeOff: .eyeOff
        case .bell: .bell
        }
    }
}

// MARK: - Facts grid (fresh state)

private struct MemoryFactsCard: View {
    let facts: [MemoryFact]
    let onOpenThread: (@MainActor () -> Void)?

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s0) {
            Text("The story behind it")
                .pantopusTextStyle(.overline)
                .foregroundStyle(Theme.Color.appTextSecondary)
                .padding(.horizontal, Spacing.s3)
                .padding(.vertical, Spacing.s2)
                .frame(maxWidth: .infinity, alignment: .leading)
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)

            ForEach(Array(facts.enumerated()), id: \.element.id) { index, fact in
                factEntry(fact)
                if index < facts.count - 1 {
                    Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
                }
            }
        }
        .background(Theme.Color.appSurface)
        .clipShape(RoundedRectangle(cornerRadius: Radii.xl))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.xl)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
    }

    @ViewBuilder private func factEntry(_ fact: MemoryFact) -> some View {
        if fact.linkHint != nil, let onOpenThread {
            Button(action: { onOpenThread() }) {
                factRow(fact, showHint: true)
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("memoryBody.openThread")
            .accessibilityHint("Opens the original Pulse thread")
        } else {
            factRow(fact, showHint: false)
        }
    }

    private func factRow(_ fact: MemoryFact, showHint: Bool) -> some View {
        HStack(alignment: .top, spacing: Spacing.s3) {
            ZStack {
                RoundedRectangle(cornerRadius: Radii.sm).fill(Theme.Color.warningBg)
                Icon(icon(for: fact.kind), size: 13, color: Theme.Color.warning)
            }
            .frame(width: 24, height: 24)

            VStack(alignment: .leading, spacing: 1) {
                Text(fact.label)
                    .font(.system(size: 11, weight: .semibold))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                Text(fact.value)
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.Color.appText)
                    .fixedSize(horizontal: false, vertical: true)
                if showHint, let hint = fact.linkHint {
                    Text(hint)
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(Theme.Color.primary600)
                }
            }
            Spacer(minLength: Spacing.s0)
        }
        .frame(maxWidth: .infinity, minHeight: 44, alignment: .leading)
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, Spacing.s2)
        .contentShape(Rectangle())
    }

    private func icon(for kind: MemoryFact.Kind) -> PantopusIcon {
        switch kind {
        case .anniversary: .calendar
        case .pulseThread: .messageSquare
        case .location: .mapPin
        case .others: .users
        }
    }
}

// MARK: - Vault location card (saved state)

private struct MemoryVaultCard: View {
    let vault: MemoryVaultInfo
    let onOpenVault: (@MainActor () -> Void)?

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s3) {
            Text("Filed in your Vault")
                .pantopusTextStyle(.overline)
                .foregroundStyle(Theme.Color.appTextSecondary)

            breadcrumb
            stats
            if let onOpenVault {
                openButton(onOpenVault)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Spacing.s4)
        .background(Theme.Color.appSurface)
        .clipShape(RoundedRectangle(cornerRadius: Radii.xl))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.xl)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
    }

    private var breadcrumb: some View {
        ScrollView(.horizontal) {
            HStack(spacing: Spacing.s1) {
                ForEach(Array(vault.trail.enumerated()), id: \.element.id) { index, crumb in
                    crumbChip(crumb)
                    if index < vault.trail.count - 1 {
                        Icon(.chevronRight, size: 11, color: Theme.Color.appTextMuted)
                    }
                }
            }
            .padding(.horizontal, Spacing.s2)
            .padding(.vertical, Spacing.s2)
        }
        .scrollIndicators(.hidden)
        .background(Theme.Color.appSurfaceSunken)
        .clipShape(RoundedRectangle(cornerRadius: Radii.md))
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Path: " + vault.trail.map(\.label).joined(separator: ", "))
    }

    private func crumbChip(_ crumb: MemoryVaultCrumb) -> some View {
        HStack(spacing: Spacing.s1) {
            Icon(icon(for: crumb.glyph), size: 11, color: crumb.isCurrent ? Theme.Color.appText : Theme.Color.appTextStrong)
            Text(crumb.label)
                .font(.system(size: 12, weight: crumb.isCurrent ? .bold : .semibold))
                .foregroundStyle(crumb.isCurrent ? Theme.Color.appText : Theme.Color.appTextStrong)
        }
        .padding(.horizontal, Spacing.s2)
        .padding(.vertical, Spacing.s1)
        .background(crumb.isCurrent ? Theme.Color.appSurface : Color.clear)
        .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.pill)
                .stroke(crumb.isCurrent ? Theme.Color.appBorder : Color.clear, lineWidth: 1)
        )
    }

    private var stats: some View {
        HStack(spacing: Spacing.s0) {
            ForEach(Array(vault.stats.enumerated()), id: \.element.id) { index, stat in
                if index > 0 {
                    Rectangle().fill(Theme.Color.appBorderSubtle).frame(width: 1, height: 28)
                }
                VStack(spacing: 2) {
                    Text(stat.value)
                        .font(.system(size: 14, weight: .heavy))
                        .foregroundStyle(Theme.Color.appText)
                        .lineLimit(1)
                        .minimumScaleFactor(0.7)
                    Text(stat.label)
                        .font(.system(size: 10, weight: .semibold))
                        .foregroundStyle(Theme.Color.appTextSecondary)
                }
                .frame(maxWidth: .infinity)
            }
        }
        .padding(.vertical, Spacing.s2)
        .background(Theme.Color.appSurface)
        .clipShape(RoundedRectangle(cornerRadius: Radii.md))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.md)
                .stroke(Theme.Color.appBorderSubtle, lineWidth: 1)
        )
    }

    private func openButton(_ action: @escaping @MainActor () -> Void) -> some View {
        Button(action: { action() }) {
            HStack(spacing: Spacing.s1) {
                Text("Open Vault › Memories")
                    .font(.system(size: 13, weight: .bold))
                    .foregroundStyle(Theme.Color.appText)
                Icon(.arrowRight, size: 13, color: Theme.Color.appText)
            }
            .frame(maxWidth: .infinity, minHeight: 44)
            .background(Theme.Color.appSurface)
            .clipShape(RoundedRectangle(cornerRadius: Radii.md))
            .overlay(
                RoundedRectangle(cornerRadius: Radii.md)
                    .stroke(Theme.Color.appBorder, lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("memoryBody.openVault")
        .accessibilityLabel("Open Vault, Memories")
    }

    private func icon(for glyph: MemoryVaultCrumb.Glyph) -> PantopusIcon {
        switch glyph {
        case .inbox: .inbox
        case .archive: .archive
        case .heart: .heart
        case .calendar: .calendar
        }
    }
}

#Preview("Fresh") {
    let onOpenThread: @MainActor () -> Void = {}
    return ScrollView {
        MemoryBody(memory: MemorySampleData.memory, isSaved: false, onOpenThread: onOpenThread)
            .padding(.vertical)
    }
    .background(Theme.Color.appBg)
}

#Preview("Saved") {
    let onOpenVault: @MainActor () -> Void = {}
    return ScrollView {
        MemoryBody(memory: MemorySampleData.savedMemory, isSaved: true, onOpenVault: onOpenVault)
            .padding(.vertical)
    }
    .background(Theme.Color.appBg)
}
